src/util/prov_helpers.rs

```diff
@@ -1,39 +1,279 @@
+            // The parent record must also use the address being bound to as its address in order for
+            // the bind to succeed.  This is the only way in which Provenance accepts a non-restricted
+            // name bind
+            address: bind_address,
+            restricted: false,
+        })
+    } else {
+        None
+    };
+    MsgBindNameRequest {
+        record: bind_record,
+        parent: parent_record,
+    }
+    .to_ok()
+}
+
+fn build_page_request(key: Vec<u8>) -> Option<PageRequest> {
+    Some(PageRequest {
+        key: key,
+        offset: 0,
+        limit: 100,
+        count_total: false,
+        reverse: false,
+    })
+}
+
+/// Fetches all attributes for an address, handling paging as appropriate
+/// 
+/// # Parameters
+/// * `querier` The Provenance Blockchain AttributeQuerier to use for fetching pages of attributes
+/// * `address` The address to fetch attributes on
+pub fn get_all_attributes<Q: CustomQuery, S1: Into<String> + Copy>(
+    querier: AttributeQuerier<Q>,
+    address: S1,
+) -> Result<Vec<QueryAttributesResponse>, ContractError> {
+    let mut results = vec![];
+    let mut pagination = build_page_request(vec![]);
+    loop {
+        let res = querier.attributes(address.into(), pagination.clone())?;
+        results.push(res.clone());
+        match res.pagination {
+            Some(PageResponse {
+                next_key: Some(..), ..
+            }) => pagination = build_page_request(res.pagination.unwrap().next_key.unwrap()),
+            Some(PageResponse { next_key: None, .. }) => break,
+            None => break,
+        }
+    }
+
+    Ok(results)
+}
+
 #[cfg(test)]
 mod tests {
-    use crate::util::prov_helpers::get_group_id_attribute_values;
-    use cosmwasm_std::{to_binary, Addr, Binary};
-    use provwasm_std::{Attribute, AttributeValueType, Attributes};
+    use crate::{
+        test::test_constants::{DEFAULT_CONTRACT_ATTRIBUTE, DEFAULT_GROUP_MEMBER},
+        types::core::error::ContractError,
+        util::prov_helpers::{get_group_id_attribute_values, msg_bind_name},
+    };
+    use cosmwasm_std::{from_json, to_json_vec, Binary, ContractResult, SystemResult};
+    use provwasm_mocks::{mock_provenance_dependencies, MockProvenanceQuerier};
+    use provwasm_std::types::{
+        cosmos::base::query::v1beta1::{PageRequest, PageResponse},
+        provenance::attribute::v1::{
+            Attribute, AttributeQuerier, AttributeType, QueryAttributesRequest,
+            QueryAttributesResponse,
+        },
+        tendermint::abci::ResponseQuery,
+    };
+
+    use super::get_all_attributes;
+
+    /// Helper function to handle returning mocked responses with pagination, where the key is a json-encoded string of the next page index
+    fn mock_paged_attributes(
+        querier: &mut MockProvenanceQuerier,
+        account: &'static str,
+        pages: Vec<Vec<Attribute>>,
+    ) {
+        querier.registered_custom_queries.insert(
```

Comment:
turns out this is just how complicated it is to have dynamic handling of multiple queries to the mock querier (vs. the simplistic `QueryAttributesRequest::mock_response(` type single-response-value mock)